Signal/src/Models/TSMessageAdapaters/AttachmentUploadView.m

```diff
@@ -0,0 +1,114 @@
+
+@end
+
+#pragma mark -
+
+@implementation AttachmentUploadView
+
+- (instancetype)initWithAttachment:(TSAttachmentStream *)attachment
+                         superview:(UIView *)superview
+           attachmentStateCallback:(AttachmentStateBlock)attachmentStateCallback
+{
+    self = [super init];
+
+    if (self) {
+        OWSAssert(attachment);
+        OWSAssert(superview);
+
+        self.attachment = attachment;
+        self.attachmentStateCallback = attachmentStateCallback;
+
+        _maskLayer = [CALayer layer];
+        [_maskLayer setBackgroundColor:[UIColor blackColor].CGColor];
+        [_maskLayer setOpacity:0.4f];
+        [_maskLayer setFrame:superview.frame];
+        [superview.layer addSublayer:_maskLayer];
+
+        const CGFloat progressWidth = round(superview.frame.size.width * 0.45f);
+        const CGFloat progressHeight = round(progressWidth * 0.11f);
+        CGRect progressFrame = CGRectMake(round((superview.frame.size.width - progressWidth) * 0.5f),
+            round((superview.frame.size.height - progressHeight) * 0.5f),
+            progressWidth,
+            progressHeight);
+        _progressView = [OWSProgressView new];
+        _progressView.frame = progressFrame;
+        [superview addSubview:_progressView];
+
+        [[NSNotificationCenter defaultCenter] addObserver:self
+                                                 selector:@selector(attachmentUploadProgress:)
+                                                     name:kAttachmentUploadProgressNotification
+                                                   object:nil];
+
+        _isAttachmentReady = self.attachment.isUploaded;
+
+        [self ensureViewState];
+
+        if (attachmentStateCallback) {
+            self.attachmentStateCallback(_isAttachmentReady);
+        }
+    }
+    return self;
+}
+
+- (void)dealloc
+{
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)setIsAttachmentReady:(BOOL)isAttachmentReady
+{
+    if (_isAttachmentReady == isAttachmentReady) {
+        return;
+    }
+
+    _isAttachmentReady = isAttachmentReady;
+
+    [self ensureViewState];
+
+    if (self.attachmentStateCallback) {
+        self.attachmentStateCallback(isAttachmentReady);
+    }
+}
+
+- (void)ensureViewState
+{
+    _maskLayer.hidden = self.isAttachmentReady;
+    _progressView.hidden = self.isAttachmentReady;
+}
+
+- (void)attachmentUploadProgress:(NSNotification *)notification
+{
+    NSDictionary *userinfo = [notification userInfo];
+    double progress = [[userinfo objectForKey:kAttachmentUploadProgressKey] doubleValue];
+    NSString *attachmentID = [userinfo objectForKey:kAttachmentUploadAttachmentIDKey];
+    if ([self.attachment.uniqueId isEqualToString:attachmentID]) {
+        if (!isnan(progress)) {
+            [_progressView setProgress:(float)progress];
+            self.isAttachmentReady = self.attachment.isUploaded;
+        } else {
+            self.isAttachmentReady = YES;
```

Comment:
I guess, although the bug would have to be in AFNetworking.  I wouldn't have been paranoid enough to add this check myself, but since it was in the progress logic that I used as my point of my departure, I decided to leave it in. I've added an assert.